### PR TITLE
add babel and separate harmony, script, module modes

### PR DESCRIPTION
**Pull request by caub**
- more options: 
  - `n>`: classic stable node-cjs
  - `h>`: idem with --harmony and --experimental flags
  - `b>`: babel stage1+ compilation
  - `s>`: uses vm.Script (so no process, global or require object available)
  - `m>` uses vm.SourceTextModule

- the Dockerfile directly shipped in, it makes things a bit easier, we could later for example `npm i moment` or whatever in it, to expose that module (as a separate Dockerfile to avoid polluting others).

- removed yarn.lock: I needed to use `npm ls ..` for solving issues, and npm is faster on my OS, if it matters, I can put back yarn now

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -1,5 +1,8 @@
 {
   "plugins": [
+    // Stage 0
+    "@babel/plugin-proposal-function-bind",
```
Comment (ljharb) [inline .babelrc]: I wouldn’t recommend including this one; it’s unlikely to progress at this time.

File: src/plugins/js-eval/jsEvalPlugin.js [hunk 2]
```diff
@@ -1,56 +1,79 @@
...
 };
 
 if (process.env.NODE_ENV !== 'test') {
   process.on('exit', () => {
-    for (const child of children) {
-      child.kill();
-    }
+    cp.exec('docker rm -f $(docker ps -qf name=jseval)', (err, stdout) => {
+      console.log(stdout);
+    });
```
Comment (brigand) [inline src/plugins/js-eval/jsEvalPlugin.js]: Something like this...

```js
const cmd = 'docker rm -f $(docker ps -qf name=jseval)';
cp.exec(cmd (err, stdout) => {
  console.error(`Command '${cmd}' finished with output`);
  console.error(stdout);
  console.error(`End of '${cmd}' output`);
});
```

File: scripts/deploy.js [hunk 3]
```diff
@@ -80,7 +80,7 @@ async function run() {
       console.error(`Run with --restart to restart the bot.`);
     }
 
-    await ssh.exec(`docker pull brigand/js-eval`);
+    await ssh.exec(`node src/plugins/js-eval/init`); // build brigand/js-eval image
```
Comment (brigand) [inline scripts/deploy.js]: How long does this take to execute if the image is unchanged?

File: src/plugins/js-eval/__tests__/jsEvalPlugin-test.js [hunk 4]
```diff
@@ -1,60 +1,108 @@
...
   });
 
-  it(`errors when it should`, () => {
-    jsEval({
+  it(`errors when it should`, async () => {
+    await jsEval({
       message: 'n> 2++2',
       respond: output => {
-        expect(output).toEqual(`(error) ecmabot.js:1
+        expect(output).toEqual(`Error: ecmabot.js:1
```
Comment (brigand) [inline src/plugins/js-eval/__tests__/jsEvalPlugin-test.js]: Can we remove this `ecmabot.js:1` stuff? Can be a separate PR.

File: src/plugins/js-eval/init.js [hunk 5]
```diff
@@ -1,42 +1,32 @@
...
 
 async function initJsEval() {
   // Kill all docker containers created with "--name jseval-{some hex string}"
-  exec(`docker ps --format '{{ json . }}'`, { encoding: 'utf-8'}, (err, stdout = '') => {
-    const containers = stdout.split(/[\r\n]+/)
-      .map(x => x.trim())
-      .filter(Boolean)
-      .map(tryParse)
-      .filter(Boolean);
+  const stdout = await exec(`docker rm -f $(docker ps -aqf name=jseval-)`).catch(() => ''); // ignore if no images
```
Comment (brigand) [inline src/plugins/js-eval/init.js]: Maybe .trim() for good measure.

**Comment by brigand** [comment]
Looks good, just some minor issues.

File: src/plugins/js-eval/init.js (hunk 5 above)
Comment (caub) [reply src/plugins/js-eval/init.js]: yea `exec` (line 4) does the trimming

File: scripts/deploy.js (hunk 3 above)
Comment (brigand) [reply scripts/deploy.js]: That's about 7 seconds longer than it needs to be :-p Could we try this?

- hash the docker file
- compare to Dockerfile.hash on the server (if it exists)
- if not equal, build the container and if successful write the hash to Dockerfile.hash

File: src/plugins/js-eval/__tests__/jsEvalPlugin-test.js (hunk 4 above)
Comment (caub) [reply src/plugins/js-eval/__tests__/jsEvalPlugin-test.js]: Hmm without [FILENAME](https://github.com/devsnek/docker-js-eval/blob/master/run.js#L15) 
```
Error: ecmabot.js:1\n1 ++ 1\n^\n\nReferenceError: Invalid left-hand side expression in postfix operation
```
changes to
```
Error: evalmachine.<anonymous>:1\n1 ++ 1\n^\n\nReferenceError: Invalid left-hand side expression in postfix operation
```

Comment (caub) [reply src/plugins/js-eval/__tests__/jsEvalPlugin-test.js]: done, only logging error name and message now https://github.com/brigand/jellobot/pull/22/commits/1f15b60680f72c888d2c5193715188cd025b294d#diff-8c048966e903e3957716694dd66d77f6

the `   ^` stuff to indicate where is the error is interesting, but less useful when the line-returns are removed

File: scripts/deploy.js (hunk 3 above)
Comment (caub) [reply scripts/deploy.js]: hmm, a bit hard to do (above all since I've added `run.js` COPY'ed from `Dockerfile`). Does it really matter? the deploy don't need to be frequent

Note that if you can provide a `GH_TOKEN` env var during the build, it fetches the latest nodejs version automatically

Comment (brigand) [reply scripts/deploy.js]: Yeah, that's fine. Also the cron is a good idea.
